### DISCUSSION TITLE
Leaderboard not working for #68

### DISCUSSION THREAD

**Discussion by Beowolve** [cssbattle/forum#198] [Bugs reports]
The leaderboard for [target #68](https://cssbattle.dev/leaderboard/target/68) is not working (blank screen). Other targets seam to have no problem, very strange...

I suspect, that is has something to do with a missing avatar image, this is the console output:

```
Uncaught (in promise) TypeError: Cannot read properties of null (reading 'avatar')
    at LeaderboardPage.jsx:368
    at Array.map (<anonymous>)
    at Ye (LeaderboardPage.jsx:362)
    at na (react-dom.production.min.js:157)
    at Fa (react-dom.production.min.js:180)
    at qs (react-dom.production.min.js:269)
    at Tu (react-dom.production.min.js:250)
    at _u (react-dom.production.min.js:250)
    at wu (react-dom.production.min.js:250)
    at hu (react-dom.production.min.js:243)
    at react-dom.production.min.js:123
    at e.unstable_runWithPriority (scheduler.production.min.js:18)
    at qi (react-dom.production.min.js:122)
    at Gi (react-dom.production.min.js:123)
    at zi (react-dom.production.min.js:122)
    at su (react-dom.production.min.js:237)
    at Oa (react-dom.production.min.js:170)
    at LeaderboardPage.jsx:106
 ```

**Comment by SelenIT** [comment]
It seems to be missing the whole user data for a player with userId=`IRk8FW9kKkdx16ogoCIh3YOA92j1` who got 90th place with 224 characters:
![image](https://user-images.githubusercontent.com/282754/145795692-3155474d-ebeb-4537-9cad-1876e8915f5c.png)

**Comment by chinchang** [comment]
We'll have a look at this. Thanks for reporting!

**Reply by Beowolve** [reply]
> We'll have a look at this. Thanks for reporting!

Thanks for looking at it, I solved in the meantime by making a top 10 solution 😎

**Comment by Beowolve** [comment]
[Taget#95](https://cssbattle.dev/leaderboard/target/95) now has the same problem. Looks like a user was deleted and there is a NULL reference to his profile now.

I see two options:
* Check for NULLs and/or replace user profile with a dummy "deleted user" profile.
* Remove records of deleted users

Personally, I would implement option two, there is no point for an entry in the leaderboard if the profile is deleted.
I don't know which backend database is used, but for standard SQL databases, foreign key constraints with cascade delete would be the standard way of solving this. NoSQL usually does not have such concepts, so I fear u need to clean up your references with some extra code if a user gets deleted in that case.

**Reply by chinchang** [reply]
FIxed this, once and for all.

**Reply by gunnarbittersmann** [reply]
I disagree with @Beowolve. IMHO, submitted solutions (esp. top-ten solutions) should stay on the leaderboards after their submitters have deleted their accounts. The player’s name might read “anonymous” or similar then.
But of course, the leaderboard shouldn’t crash in case of missing data.

**Reply by Beowolve** [reply]
I see your point, but even for top 10 solutions, I disagree for the all-time leaderboard. You can leave your account inactive if you want your records still visible, but a leaderboard with a bunch of anonymous wouldn't be fun anymore. Deleted should mean deleted...
For individual battle leaderboards from the past, it's different. There, it makes sense to leave anonymous entries to keep the top 100 list consistent. 

Anyway, it's just a rare border case that in reality isn't going to happen for top 10 solutions very often.

**Reply by gunnarbittersmann** [reply]
That’s just not true. Should [Viacheslav Popov](https://cssbattle.dev/player/anwerso) delete his profile, we would lose the history of all early targets. The top-tens of targets 2, 3, 4, 5, 6, 7 (🥇), 8, 9 (🥈), 10 (🥉), 11, 12, 14, 15 (🥈), 16, 17 (🥉), 18, 19, 20, 21, 22, 23, 24 (🥉), 25 (🥈), 26 (🥇), 27, 28, 30 (🥉), 31, 32, 33, 34, 35 (🥇), 36, 37 (🥉), 39, 41 (🥈), 42, 43, 44 (🥈), 45, and 46 would be altered.

I hope @chinchang chose option 1 for the fix.

**Reply by chinchang** [reply]
We have just handled the null users for now. No records are being deleted. Infact, these users have not deleted profiles. It seems some data from their profile is missing. We'll investigate that further.
